client/src/reports/createlinksproxymodel.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+  This file is part of FatCRM, a desktop application for SugarCRM written by KDAB.
+
+  Copyright (C) 2019 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+  Authors: Kevin Funk <kevin.funk@kdab.com>
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "createlinksproxymodel.h"
+
+#include "accountdataextractor.h"
+#include "contactdataextractor.h"
+#include "entitytreemodel.h"
+#include "itemstreemodel.h"
+#include "opportunitydataextractor.h"
+
+using namespace Akonadi;
+
+CreateLinksProxyModel::CreateLinksProxyModel(const QString &resourceBaseUrl, QObject* parent)
+    : QIdentityProxyModel(parent)
+    , mResourceBaseUrl(resourceBaseUrl)
+{
+}
+
+void CreateLinksProxyModel::setSourceModel(QAbstractItemModel* sourceModel)
+{
+    QIdentityProxyModel::setSourceModel(sourceModel);
+
+    // find the underlying ItemsTreeModel
+    auto model = sourceModel;
+    do {
+        if (auto *itemsTreeModel = qobject_cast<ItemsTreeModel *>(model)) {
+            mItemsTreeModel = itemsTreeModel;
+            break;
+        }
+    } while ((model = qobject_cast<QAbstractProxyModel *>(model)->sourceModel()));
+    Q_ASSERT(mItemsTreeModel);
+}
+
+QVariant CreateLinksProxyModel::data(const QModelIndex& index, int role) const
+{
+    if (!mItemsTreeModel) {
+        // should not happen
+        return QIdentityProxyModel::data(index, role);
+    }
+
+    if ((role == Qt::DisplayRole) || (role == Qt::EditRole)) {
+        auto linkedName = [this, role](const ItemDataExtractor &itemDataExtractor, const QModelIndex &index) {
+            const Item item = index.data(EntityTreeModel::ItemRole).value<Item>();
+            const QString name = sourceModel()->data(mapToSource(index), role).toString();
+            const QUrl url = itemDataExtractor.itemUrl(mResourceBaseUrl, item);
+            if (url.isValid())
+                return QStringLiteral("<html><a href=\"%1\">%2</a></html>").arg(url.toString(), name );
+
+            return name;
+        };
+
+        const int column = index.column();
+
+        switch (mItemsTreeModel->columnTypes().at(column)) {
```

Comment:
This line of code will break if any proxymodel in between, plays with the columns.
(it's passing a "proxy column" to the source model, there could be a QSFPM or RearrangeColumnsProxyModel in between that makes this invalid).
You're lucky that in your case the RearrangeColumns Proxy is on top of this one and not the other way around ;-)

I think it would be more future-proof to mapToSource(index) all the way down to mItemsTreeModel (which requires a loop). Or better, adding a ColumnTypeRole in ItemsTreeModel, then your data() call to get the column type will traverse proxies transparently.

Reply:
> You're lucky that in your case the RearrangeColumns Proxy is on top of this one and not the other way around ;-)

Lucky? I did that intentionally :)

Let me see if I can make this more fail-safe (actually, I didn't want to spend too much time on it)  :)